sympy/polys/rings.py

```diff
@@ -3005,3 +3027,243 @@ def sqf_list(f, all=False):
 
     def factor_list(f):
         return f.ring.dmp_factor_list(f)
+
+def monomial_extract(p):
+    """
+    Extracts any common monomial from the polynomials in p.
+
+    Examples
+    ========
+
+    >>> from sympy import ZZ, ring, monomial_extract
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = x + y
+    >>> g = g = x**2*y + x*y
+    >>> p = [f, g]
+    >>> monomial_extract(p)
+    ([x + y, x**2*y + x*y], None)
+
+    >>> f = x**2*y
+    >>> g = x**2*y + x*y
+    >>> p = [f, g]
+    >>> monomial_extract(p)
+    ([x**2*y, x**2*y + x*y], x*y)
+
+    """
+    ring = p[0].ring
+    zero_monom = ring.zero_monom
+
+    monomials = chain(*p)
+    monom_gcd = tuple(map(min, zip(*monomials)))
+
+    # If the zero monomial is present, we can immediately return
+    if monom_gcd == zero_monom:
+        return p, None
+    else:
+        # Loop over a few monomials to quickly check if gcd is zero
+        for i, monom in enumerate(monomials):
+            if i >= 2:
+                break
+            gcd_monom = tuple(map(min, zip(monom_gcd, monom)))
+            if gcd_monom == zero_monom:
+                return p, None
+
+        d = ring({monom_gcd: ring.domain.one})
+        _d = degree_list(d.as_expr())
+        print(_d)
+        p_result = []
+        for pi in p:
+            pi = degree_list(pi.as_expr())
+            print(pi)
+            p_result.append(ring.monomial_ldiv(pi, _d))
+        p_result = p
+        return p, d
+
+def gcd_ground(coeffs, domain):
+    """
+    Compute the greatest common divisor (GCD) of elements in the ground domain.
```

Comment:
There already is a function that does this:
https://github.com/sympy/sympy/blob/feeccac64b70b9df6c5401c55b98a35f6e07ed8c/sympy/polys/densetools.py#L571-L610
I'm not sure exactly why `QQ` is special-cased there. I guess there is some sort of definition for what gcd in QQ is supposed to be.

There is also a method on `PolyElement` although perhaps it could be improved to be more like `dup_content` (it does not shortcut):
https://github.com/sympy/sympy/blob/feeccac64b70b9df6c5401c55b98a35f6e07ed8c/sympy/polys/rings.py#L1956-L1965